src/renderer/viz/expressions/top.js

```diff
@@ -59,56 +67,87 @@ export default class Top extends BaseExpression {
         this._textureBuckets = null;
     }
     get numCategories () {
-        return Math.round(this.buckets.eval()) + 1;
+        return this.numBuckets + 1;
+    }
+    get numBuckets () {
+        let buckets = Math.round(this.buckets.eval());
+        if (buckets > this.property.numCategories) {
+            buckets = this.property.numCategories;
+        }
+        if (buckets > MAX_TOP_BUCKETS) {
+            // setTimeout is used here because throwing within the renderer stack leaves the state in an invalid state,
+            // making this error an unrecoverable error, within the setTimeout the error is recoverable
+            const prev = this.buckets.eval();
```

Comment:
Why not using `buckets`:
```js
const prev = buckets;
```

Reply:
Because `buckets` has been rounded and clamped